### DISCUSSION TITLE
Integrating NextJS with AdminJS

### DISCUSSION THREAD

**Discussion by cirosayagueslaso** [vercel/community#4488] [Help]
### Summary

Hey there!

I'm trying to integrate [AdminJS](https://docs.adminjs.co/) into a Next application and deploy it to Vercel.

I'm able to run it locally and have the Next app in the base URL and the AdminJS in `/xyz-admin` but when I deploy it to Vercel I get some strange error 500 which I can't understand or explain, this is the first line of what I see in the deployment log on Vercel: "Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/var/task/node_modules/adminjs/lib/locale/de/translation.json' imported from /var/task/node_modules/adminjs/lib/locale/index.js".

I've researched for hours and tried several things before starting this discussion but I couldn't find the correct way to solve this.

In case someone can and wants to take a look: [here](https://github.com/cirosayagueslaso/adminjs-nextjs) is my repository.

This is the entire log associated to the error 500:

```
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/var/task/node_modules/adminjs/lib/locale/de/translation.json' imported from /var/task/node_modules/adminjs/lib/locale/index.js
    at new NodeError (node:internal/errors:405:5)
    at finalizeResolution (node:internal/modules/esm/resolve:329:11)
    at moduleResolve (node:internal/modules/esm/resolve:992:10)
    at moduleResolveWithNodePath (node:internal/modules/esm/resolve:936:12)
    at defaultResolve (node:internal/modules/esm/resolve:1178:79)
    at nextResolve (node:internal/modules/esm/loader:163:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:835:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:424:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:77:40)
    at link (node:internal/modules/esm/module_job:76:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
MissingStaticPage [Error]: Failed to load static file for page: /en/500 ENOENT: no such file or directory, open '/var/task/.next/server/pages/en/500.html'
    at /var/task/node_modules/next/dist/server/require.js:109:19
    at async loadComponentsImpl (/var/task/node_modules/next/dist/server/load-components.js:80:26)
    at async NextNodeServer.findPageComponentsImpl (/var/task/node_modules/next/dist/server/next-server.js:439:36)
    at async NextNodeServer.renderErrorToResponseImpl (/var/task/node_modules/next/dist/server/base-server.js:1651:30)
    at async pipe.req.req (/var/task/node_modules/next/dist/server/base-server.js:1591:30)
    at async NextNodeServer.pipeImpl (/var/task/node_modules/next/dist/server/base-server.js:735:25)
    at async NextNodeServer.handleCatchallRenderRequest (/var/task/node_modules/next/dist/server/next-server.js:729:17)
    at async NextNodeServer.runImpl (/var/task/node_modules/next/dist/server/base-server.js:720:9)
    at async NextNodeServer.handleRequestImpl (/var/task/node_modules/next/dist/server/base-server.js:668:20)
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/var/task/node_modules/adminjs/lib/locale/de/translation.json' imported from /var/task/node_modules/adminjs/lib/locale/index.js
    at new NodeError (node:internal/errors:405:5)
    at finalizeResolution (node:internal/modules/esm/resolve:329:11)
    at moduleResolve (node:internal/modules/esm/resolve:992:10)
    at moduleResolveWithNodePath (node:internal/modules/esm/resolve:936:12)
    at defaultResolve (node:internal/modules/esm/resolve:1178:79)
    at nextResolve (node:internal/modules/esm/loader:163:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:835:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:424:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:77:40)
    at link (node:internal/modules/esm/module_job:76:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

### Example

https://adminjs-nextjs-q40ogmpj4-ciros-projects-2496ab1f.vercel.app/xyz-admin

### Steps to Reproduce

1. Import the repo to a github account
3. Deploy it to Vercel by connecting the imported project from Vercel panel.
4. Access the generated site's admin section: [url]/xyz-admin

**Comment by abhishekyelley** [comment]
Hello @cirosayagueslaso, I realize that it's been a couple of months since you posted this question. I faced the exact same problem aswell and I thought I'd write this so anyone coming here in the future might have a better time figuring out things.

# 1
Vercel seems to have a problem when dealing with imports assertions. I'm not exactly sure why but maybe [this](https://v8.dev/features/import-assertions#background%3A-json-modules-and-mime-type) or [this](https://github.com/tc39/proposal-import-attributes?tab=readme-ov-file#history) might help.

Inside `node_modules/adminjs/lib/locale/index.js`, you would find this:
```javascript
import deLocale from './de/translation.json' assert { type: 'json' };
import enLocale from './en/translation.json' assert { type: 'json' };
import esLocale from './es/translation.json' assert { type: 'json' };
...
```

I had to replace the keyword `assert` with `with` like so:
```javascript
import deLocale from './de/translation.json' with { type: 'json' };
```

**NOTE: This only works with node v20.x**

I used [patch-package](https://www.npmjs.com/package/patch-package) to get these changes from local to vercel. Doing this helped me get rid of that particular error. But there are some more things to take care of before getting your AdminJS app working on vercel.

# 2
The above step has fixed the error you talk about in your post.However, I have recieved a new error in vercel regarding a missing module `@rollup/rollup-linux-x64-gnu`. I was unable to find out why this happened and was not able to replicate this error on all deployments. What I did to fix it was to delete my `node_modules` and `package-lock.json` and install `@rollup/rollup-linux-x64-gnu` as mentioned [here](https://answers.netlify.com/t/react-vite-build-fails-because-of-rollup-rollup-linux-x64-gnu/107504/4) ,
OR
Add to `package.json` the following:
```JSON
"overrides": {
    "rollup": "npm:@rollup/wasm-node"
}
```
as mentioned [here](https://github.com/vitejs/vite/issues/15167).

# 3
Make sure to pre-bundle adminjs manually or by using `@adminjs/bundler` and set your environment variables: `ADMIN_JS_SKIP_BUNDLE="true"` and `NODE_ENV="production"`.
```javascript
import { bundle } from '@adminjs/bundler';
import { ComponentLoader } from 'adminjs';
const componentLoader = new ComponentLoader();

(async () => {
    const files = await bundle({
        componentLoader,
        destinationDir: 'public', // relative to CWD
    });
})();
```
Follow this guide for more details: [adminjs pre-bundling](https://adminjs-docs.web.app/module-@adminjs_bundler.html). Make sure to host the outputs on a CDN. Since you are using express, you can use:
```javascript
app.use(express.static(bundle_path))
```
and pass in the full URL to `assetsCDN` in adminOptions:
```javascript
const adminJs = new AdminJS({ assetsCDN: <your storage bucket url> })
```
Add this and patch-package to your build script in `package.json`.

# 4
Since you are using express router, you might need to rewrite all requests to that express router i.e. the folder where `@adminjs/express` app is exported. Please read this guide: [vercel-express app](https://vercel.com/guides/using-express-with-vercel)
I suspect that you might face some issues when trying to get NextJS working along with ExpressJS. You might have to completely rely on express for all the routes.

***
That should be all that you need to know to get AdminJS up and running in Vercel. Connecting AdminJS to your database will lead you into other issues where you might have to learn more about connection caching in serverless computing. [Caching connections in mongodb](https://www.youtube.com/watch?v=JIlYroSsInU).

I have a very barebones AdminJS project that is up and running on Vercel.
Github Repo: https://github.com/abhishekyelley/raw-admin
Vercel Deployment: https://raw-admin-cokt.vercel.app/admin

Hope this helps!

**Reply by cirosayagueslaso** [reply]
Hey Abhishek!! Thank you so much for the detailed answer, for sure it helps a lot!

Hope you have a great day!

Thanks!

**Reply by yourcodingguy** [reply]
hi,

 after deploying to vercel i am getting following error

Legacy server listening...
Unhandled Promise Rejection     {"errorType":"Runtime.UnhandledPromiseRejection","errorMessage":"Error: ENOENT: no such file or directory, mkdir '.adminjs'","reason":{"errorType":"Error","errorMessage":"ENOENT: no such file or directory, mkdir '.adminjs'","code":"ENOENT","errno":-2,"syscall":"mkdir","path":".adminjs","stack":["Error: ENOENT: no such file or directory, mkdir '.adminjs'"]},"promise":{},"stack":["Runtime.UnhandledPromiseRejection: Error: ENOENT: no such file or directory, mkdir '.adminjs'","    at process.<anonymous> (file:///var/runtime/index.mjs:1276:17)","    at process.emit (node:events:530:35)","    at emit (node:internal/process/promises:150:20)","    at processPromiseRejections (node:internal/process/promises:284:27)","    at process.processTicksAndRejections (node:internal/process/task_queues:96:32)"]}
Unhandled rejection: [Error: ENOENT: no such file or directory, mkdir '.adminjs'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'mkdir',
  path: '.adminjs'
}
Error: Runtime exited with error: exit status 1
Runtime.ExitError

**Reply by abhishekyelley** [reply]
Hey @yourcodingguy, could you share your code or provide me with a brief summary?
